src/Microsoft.Extensions.Logging.Console/Internal/ConsoleLoggerProcessor.cs

```diff
@@ -0,0 +1,109 @@
+﻿// Copyright (c) .NET Foundation. All rights reserved.
+// Licensed under the Apache License, Version 2.0. See License.txt in the project root for license information.
+
+using System;
+using System.Collections.Concurrent;
+using System.Threading;
+using System.Threading.Tasks;
+
+namespace Microsoft.Extensions.Logging.Console.Internal
+{
+    public class ConsoleLoggerProcessor
+    {
+        private const int _maxQueuedMessages = 1024;
+        // Writing to console is not an atomic operation in the current implementation and since multiple logger 
+        // instances are created with a different name. Also since Console is global, using a static lock is fine. 
+        private static readonly object _lock = new object();
+
+        private IConsole _console;
+
+        private readonly BlockingCollection<LogMessageEntry> _messageQueue = new BlockingCollection<LogMessageEntry>(_maxQueuedMessages);
+        private readonly Task _outputTask;
+
+        public ConsoleLoggerProcessor()
+        {
+            RegisterForExit();
+
+            // Start Console message queue processor
+            _outputTask = Task.Factory.StartNew(
+                ProcessLogQueue,
+                this,
+                TaskCreationOptions.LongRunning);
+        }
+
+        public IConsole Console
+        {
+            get { return _console; }
+            set
+            {
+                if (value == null)
+                {
+                    throw new ArgumentNullException(nameof(value));
+                }
+
+                _console = value;
+            }
+        }
+
+        public bool HasQueuedMessages => _messageQueue.Count > 0;
+
+        public void EnqueueMessage(LogMessageEntry message)
+        {
+            _messageQueue.Add(message);
+        }
+
+        private void ProcessLogQueue()
+        {
+            foreach (var message in _messageQueue.GetConsumingEnumerable())
+            {
+                lock (_lock)
+                {
+                    if (message.LevelString != null)
+                    {
+                        Console.Write(message.LevelString, message.LevelBackground, message.LevelForeground);
+                    }
+
+                    Console.Write(message.Message, message.MessageColor, message.MessageColor);
+                    Console.Flush();
+                }
+            }
+        }
+
+        private static void ProcessLogQueue(object state)
+        {
+            var consoleLogger = (ConsoleLoggerProcessor)state;
+
+            consoleLogger.ProcessLogQueue();
+        }
+
+        private void RegisterForExit()
```

Comment:
I'd remove all of this and just make this disposable. Then call that from the console logger's dispose. Logger providers are disposed "at the right time" when wired up correctly.